Update compatibility.md: mention differing query parameter handling (#1262)
encode/httpx: docs/compatibility.md

# Requests Compatibility Guide

> HTTPX aims to be compatible with the `requests` API where we feel it makes sense.

Reviewer: I know that probably wasn't the intent, but for some reason this phrasing sounds a bit pedantic — like "we" consider that Requests' API was nonsensical in places (when really it's just a result of pioneering the field and having to live on with early decisions, which is understandable).

I also thought of replacing with "wherever sensible", but that seems to carry the same possible misunderstanding.

So I'd be keen to keep the current wording…

```suggestion
HTTPX aims to be compatible with the `requests` API where possible.
```

> HTTPX aims to be compatible with the `requests` API where we feel it makes sense.

Author: How about

```suggestion
HTTPX aims to be broadly compatible with the `requests` API.
```

I feel like the current wording creates the impression that HTTPX tries to be compatible unless that's not possible at all.

Reviewer 2: "Broadly compatible" is spot-on, yup.

`requests` defers most of its HTTP networking code to the excellent [`urllib3` library](https://urllib3.readthedocs.io/en/latest/).

On the other hand, HTTPX uses [HTTPCore](https://github.com/encode/httpcore) as its core HTTP networking layer, which is a different project than `urllib3`.

## Query Parameters

> `requests` omits `params` whose values are `None` (e.g. `requests.get(..., params={"foo": None})`). HTTPX will include a query parameter without a value instead. 

Reviewer 2: I'd probably(?) rather we were raising a TypeError here, since our annotations don't include None as a valid value for items in params. It's only by accident of implementation that it happens to currently end up being treated as equivalent to `""`.

Reviewer: Makes sense — still, we can add these docs, probably just tweaking them like this:

```suggestion
`requests` omits `params` whose values are `None` (e.g. `requests.get(..., params={"foo": None})`). This is not supported by HTTPX.
```
